### PR TITLE
Automatically attach debugger to subprocess

### PR DESCRIPTION
**Pull request by xavierzwirtz**
Uses the [Visual Studio DTE](https://docs.microsoft.com/en-us/dotnet/api/envdte.dte?view=visualstudiosdk-2019) to attach to the sub process automatically. I think packaging is implemented properly, I manually packaged this and added it to another project and saw it work. The actual implementation was sourced from https://github.com/CyAScott/AppDomainAlternative/, which has an MIT license.

### CODE REVIEW

File: src/Tmds.ExecFunction/Debugger.cs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+
+namespace Tmds.Utils
+{
+
+    /// <summary>
+    /// A wrapper for a <see cref="Process"/>.
+    /// </summary>
+    public abstract class Debugger
```
Comment (tmds) [inline src/Tmds.ExecFunction/Debugger.cs:37]: We don't want to expose this class.

File: src/Tmds.ExecFunction/ExecFunction.Program.cs [hunk 2]
```diff
@@ -49,7 +49,16 @@ public static int Main(string[] args)
                 string assemblyName = args[argIdx++];
                 string typeName = args[argIdx++];
                 string methodName = args[argIdx++];
-                string[] additionalArgs = args.SubArray(3);
+                bool debuggerEnabled = Boolean.Parse(args[argIdx++]);
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.Program.cs:52]: nit: rename `debuggerEnabled` to `waitForDebugger`.

File: src/Tmds.ExecFunction/ExecFunction.Program.cs [hunk 3]
```diff
@@ -49,7 +49,16 @@ public static int Main(string[] args)
...
+                string[] additionalArgs = args.SubArray(4);
+
+                if(debuggerEnabled)
+                {
+                    // wait for signal that debugger is attached
+
+                    Console.WriteLine("waiting for debugger");
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.Program.cs:59]: Remove the `WriteLine`.

File: src/Tmds.ExecFunction/ExecFunction.cs [hunk 4]
```diff
@@ -16,87 +16,92 @@ namespace Tmds.Utils
...
             {
                 process = new Process();
 
+                if(execFunctionHostArgs == null) { 
+                    execFunctionHostArgs = _execFunctiohHostArgsDefault.Value;
+                }
+
                 ExecFunctionOptions options = new ExecFunctionOptions(process.StartInfo);
-                ConfigureProcessStartInfoForMethodInvocation(method, args, options.StartInfo);
+                var debuggerEnabled = System.Diagnostics.Debugger.IsAttached && Debugger.DebuggingSupported;
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.cs:103]: nit: rename `debuggerEnabled` to `attachDebugger`.

File: src/Tmds.ExecFunction/ExecFunction.cs [hunk 5]
```diff
@@ -143,7 +160,7 @@ private static (Process process, Task exitedTask) Start(MethodInfo method, strin
             }
         }
 
-        private static void ConfigureProcessStartInfoForMethodInvocation(MethodInfo method, string[] args, ProcessStartInfo psi)
+        private static void ConfigureProcessStartInfoForMethodInvocation(MethodInfo method, string[] args, ProcessStartInfo psi, ExecFunctionHostArgs execFunctionHostArgs, bool debuggerEnabled)
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.cs:163]: `debuggerEnabled` -> `attachDebugger`.

File: src/Tmds.ExecFunction/ExecFunction.cs [hunk 6]
```diff
@@ -163,12 +180,16 @@ private static void ConfigureProcessStartInfoForMethodInvocation(MethodInfo meth
             // If we need the host (if it exists), use it, otherwise target the console app directly.
             Type t = method.DeclaringType;
             Assembly a = t.GetTypeInfo().Assembly;
-            string programArgs = PasteArguments.Paste(new string[] { a.FullName, t.FullName, method.Name });
+            string programArgs = PasteArguments.Paste(new string[] { a.FullName, t.FullName, method.Name, debuggerEnabled.ToString() });
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.cs:183]: -> `.ToString(CultureInfo.InvariantCulture)`.

File: src/Tmds.ExecFunction/ExecFunction.cs [hunk 7]
```diff
@@ -163,12 +180,16 @@ private static void ConfigureProcessStartInfoForMethodInvocation(MethodInfo meth
...
             string functionArgs = PasteArguments.Paste(args);
-            string fullArgs = HostArguments + " " + " " + programArgs + " " + functionArgs;
+            string fullArgs = execFunctionHostArgs.HostArguments + " " + " " + programArgs + " " + functionArgs;
 
-            psi.FileName = HostFilename;
+            psi.FileName = execFunctionHostArgs.HostFilename;
             psi.Arguments = fullArgs;
+            if(debuggerEnabled)
```
Comment (tmds) [inline src/Tmds.ExecFunction/ExecFunction.cs:189]: S `RedirectStandardInput` after `configure?.Invoke(options);` so the user can't mess with it.

File: src/Tmds.ExecFunction/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -1,3 +1,3 @@
 using System.Reflection;
 
-[assembly: AssemblyKeyFile("Tmds.ExecFunction.snk")]
\ No newline at end of file
+//[assembly: AssemblyKeyFile("Tmds.ExecFunction.snk")]
```
Comment (tmds) [inline src/Tmds.ExecFunction/Properties/AssemblyInfo.cs:3]: Uncomment this one.

File: src/Tmds.ExecFunction/Tmds.ExecFunction.csproj [hunk 9]
```diff
@@ -13,4 +13,23 @@
...
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <NoWarn>1701;1702;NU5100</NoWarn>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\VsDebugger\VsDebugger.csproj">
+      <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
+      <SkipGetTargetFrameworkProperties>true</SkipGetTargetFrameworkProperties>
+      <Private>true</Private>
```
Comment (tmds) [inline src/Tmds.ExecFunction/Tmds.ExecFunction.csproj:24]: This `Private` means that a project referencing `Tmds.ExecFunction` doesn't include the `VsDebugger` executable, right?

Can you add a comment in the XML file about this.

File: src/Tmds.ExecFunction/Tmds.ExecFunction.csproj [hunk 10]
```diff
@@ -13,4 +13,23 @@
...
+    </ProjectReference>
+  </ItemGroup>
+
+  <ItemGroup Label="FilesToCopy">
+    <Content Include="Tmds.ExecFunction.targets" PackagePath="build/Tmds.ExecFunction.targets" />
```
Comment (tmds) [inline src/Tmds.ExecFunction/Tmds.ExecFunction.csproj:29]: What does the `.targets` file do?

File: src/Tmds.ExecFunction/Tmds.ExecFunction.csproj [hunk 11]
```diff
@@ -13,4 +13,23 @@
     <Copyright>Tom Deseyn</Copyright>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <NoWarn>1701;1702;NU5100</NoWarn>
```
Comment (tmds) [inline src/Tmds.ExecFunction/Tmds.ExecFunction.csproj:17]: Is this needed?

**Comment by tmds** [comment]
@xavierzwirtz thank you for working on this. I've added some feedback.

It would be nice if we can avoid including the VsDebugger executable (and its dependencies) by default.

**Comment by tmds** [comment]
> t would be nice if we can avoid including the VsDebugger executable (and its dependencies) by default.

I wonder if [Debugger.Launch](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debugger.launch?view=net-5.0) can be used instead of the `VsDebugger` executable?

**Comment by xavierzwirtz** [comment]
> I wonder if Debugger.Launch can be used instead of the VsDebugger executable?

Benefit of using VsDebugger executable is that it uses COM to instruct the already attached Visual Studio instance to connect to the new process, no user intervention needed. Debugger.Launch causes a dialog to pop, and at least in my experience that dialog does not include the currently running VS instance since that VS instance is already attached to a process.

File: src/Tmds.ExecFunction/Tmds.ExecFunction.csproj (hunk 10 above)
Comment (xavierzwirtz) [reply src/Tmds.ExecFunction/Tmds.ExecFunction.csproj:29]: The `.targets` file handles getting the VsDebugger into the bin folder.

**Comment by tmds** [comment]
> Benefit of using VsDebugger executable is that it uses COM to instruct the already attached Visual Studio instance to connect to the new process, no user intervention needed. Debugger.Launch causes a dialog to pop, and at least in my experience that dialog does not include the currently running VS instance since that VS instance is already attached to a process.

Got it.

To get back to my comment about not including this executable by default.

The nicest would be if `Tmds.ExecFunction` can have this feature without adding a separate executable (after all this is a library for executing stuff into processes), and without adding extra dependencies (`EnvDTE*`).

If that is not possible/feasible, I think we should consider having a separate page (e.g. `Tmds.ExecFunction.VsDebugger` which is `Tmds.ExecFunction` with the debug feature.

wdyt?

**Comment by tmds** [comment]
Replaced by https://github.com/tmds/Tmds.ExecFunction/pull/15.

**Pull request closed without merging**